k8s/pkg/apis/longhorn/v1beta2/objectstore.go

```diff
@@ -0,0 +1,175 @@
+package v1beta2
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ObjectStoreState string
+
+const (
+	ObjectStoreStateUnknown  = ObjectStoreState("Unknown")
+	ObjectStoreStateStarting = ObjectStoreState("Starting")
+	ObjectStoreStateRunning  = ObjectStoreState("Running")
+	ObjectStoreStateStopping = ObjectStoreState("Stopping")
+	ObjectStoreStateError    = ObjectStoreState("Error")
+)
+
+type ObjectStoreStatus struct {
+	// The state of the object store as observed by the object store controller.
+	//
+	// The object store implemets a state machine with this, beginning at
+	// "unknown". Once the object store controller detects the new object store
+	// and begins resource creation the state is transformed to "starting". The
+	// object store controller observes the resources and once all resources are
+	// up and ready the state transitions to "running".
+	// The state remains in "running" until the object store is deleted, at
+	// which point the controller will clean up the associated resources.
+	//
+	//  │ object store created
+	//  │
+	//  └──►┌───────────┐
+	//      │ unknown   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller creates resources
+	//  │
+	//  └──►┌───────────┐
+	//      │ starting  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller detects all resources ready
+	//  │
+	//  └──►┌───────────┐
+	//      │ running   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ object endpoint is marked for deletion
+	//  │
+	//  └──►┌───────────┐
+	//      │ stopping  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller waits for dependent resources to disappear before removing
+	//  │ the finalizer, thereby letting the object store be deleted
+	//
+	// +optional
+	// +default="Unknown"
+	State ObjectStoreState `json:"state"`
+
+	// A list of addresses where the S3 store is exposed.
+	//
+	// +optional
+	Endpoints []string `json:"endpoints"`
+}
+
+type ObjectStoreCredentials struct {
+	// An access key. This will be the initial access key for the administrative
+	// account.
+	//
+	// +optional
+	AccessKey string `json:"accessKey"`
+
+	// A secret key. This is the secret key corresponding to the access key above.
+	//
+	// +optional
+	SecretKey string `json:"secretKey"`
+}
```

Comment:
Since reading a K8s resource requires access to the K8s API, which may actually include access to reading the content of a Secret resource as well, I don't think this is particularly bad. Keep in mind that the data here is only used to seed the s3gw. An operator can roll out new credentials once the gateway has been deployed via the UI. Therefore if someone really wants to secure themselves from adversaries that have read access to the K8s API, they can do so.

There are a couple of possible ways to avoid having the secret data clearly visible in the ObjectStore K8s object:

1) Replace the ObjectStoreCredentials type entirely and use a K8s object reference to name a Secret object that contains the values.

This is a bit tricky as it may require a user to create such a secret and this in turn requires them to use the correct key names for the key/values in the secret. Not a huge hurdle, but has potential for user error. But this would possibly open the door to integration with secret management tooling.

2) Try to replicate what the K8s Secret resource does to hide away the secret data from easy access. This won't completely make the secret data inaccessible, but will make it hard to accidentally leak the keys

3) Since the data here is copied by the ObjectStore controller into a Secret resource at creation time we could theoretically just delete this property from the ObjectStore once it is running. But this would possibly be a headache for backups then.

It all boils down to the question of what the threat-model of Longhorn is? Do we have to protect us from in-cluster adversaries? Do we have to protect us from adversaries with partial or full access to the K8s API?

Reply:
I have to admit that I don't have a lot of experience in security stuff. 

That being said, the common parent is your first proposal:
> Replace the ObjectStoreCredentials type entirely and use a K8s object reference to name a Secret object that contains the values.

Why does this approach have a security advantage? It is because the user can setup all kind of [Role-based Access Control](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) to protect/restrict the access to secret resource. They can also setup encryption for the secret data that is being stored in ETCD. Reference: https://kubernetes.io/docs/concepts/security/secrets-good-practices/ The model thread over here including in-cluster adversaries. If they do not have enough permission (specified by RBAC) they cannot access the key in the secret even if they have access ObjectStore CRs.

Another benefit is that this opens the door to integration with secret management tooling as you already mentioned.

For the disadvantage, this approach will surely create some headaches for backup/restore. We cannot simply copy the secret data to the backup target and later retrieve it in a different cluster. Maybe we can have a setting to let the user to choose whether they want to backup the secret data when doing ObjectStore backup. 

In the end, weighing the backup/store headaches of this approach VS the insecure problem of the original approach, I don't have a strong opinion on picking a particular approach.

What do you think?
cc @l-mb @jecluis @innobead

Reply:
Good points. If the first proposed solution here is common practice, I think it makes sense to prefer it over the others.

> For the disadvantage, this approach will surely create some headaches for backup/restore. We cannot simply copy the secret data to the backup target and later retrieve it in a different cluster.

Currently, our ClusterRole in the chart requests full permissions on secrets, so it should not be a problem to read and write secret data for backup/restore operations.
I'm not sure what else we would need to do the backup/restore between two clusters.

Reply:
> Currently, our ClusterRole in the chart requests full permissions on secrets, so it should not be a problem to read and write secret data for backup/restore operations.
I'm not sure what else we would need to do the backup/restore between two clusters.

I mean I am not sure if the user is ok with Longhorn coping secret data to backup store. It is like storing plain text password on disk. So maybe we can have a setting for user to switch on/off if they want to copy secret data to backup store? By default the setting can be turned on for the user's convenient. If user switch it off, they will have to backup secret data themself

Reply:
I'd assume the backup store is encrypted somehow since otherwise we'd be pushing the unencrypted *data* to it?

Post-deployment and after setting the initial credentials, the actual user/key combo is stored internally to s3gw anyway - https://docs.ceph.com/en/latest/radosgw/admin/#user-management - and thus included in the backup/restore, no? So whether the initial seed is backed up doesn't really matter.

While Ceph's RGW (and thus by extension eventually s3gw) supports HashiCorp, KMIP, Barbican, this is only for the Server Side encryption (eventually). Only OpenStack's Keystone or LDAP would support managing the user credentials externally, which I suspect we can't rely on being available - so at this point, every s3gw instance will have it's own user DB.

Reply:
> Post-deployment and after setting the initial credentials, the actual user/key combo is stored internally to s3gw anyway - https://docs.ceph.com/en/latest/radosgw/admin/#user-management - and thus included in the backup/restore, no? So whether the initial seed is backed up doesn't really matter.

In the current implementation, the UI needs access to credentials as well, which is supplied through a K8s secret.
If this requirement disappears with the s3gw UI backend then indeed the s3gw's database is the only relevant store of credentials post-deployment.
If the s3gw UI backend doesn't solve this requirement, the LH Object Store controller will need access to a set of credentials to configure the UI.
@jecluis do you know more details about what the UI requires in this regard in the future?

Reply:
Thanks @l-mb @m-ildefons for the feedback. Combining the feedback, I think the solution would be:
1. Replace the ObjectStoreCredentials type entirely and use a K8s object reference to name a Secret object that contains the values.
2. When doing backup/restore, we also backup/restore the credential from/into secret with the same name. This is required for UI and in case the user forgets the seed password they can find it in the secret. It has no harm since we assume that the backup target is encrypted by default.